isthmus/src/main/java/io/substrait/isthmus/type/SubstraitUserDefinedType.java

```diff
@@ -0,0 +1,251 @@
+package io.substrait.isthmus.type;
+
+import com.google.protobuf.TextFormat;
+import io.substrait.type.Type;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rel.type.RelDataTypeFieldImpl;
+import org.apache.calcite.rel.type.RelDataTypeImpl;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+/**
+ * Custom Calcite {@link RelDataType} for Substrait user-defined types.
+ *
+ * <p>This type preserves all UDT metadata (URN, name, type parameters) during Calcite roundtrips.
+ * It is used when converting types without literal context. For literals, specialized subclasses
+ * provide representation-specific handling:
+ *
+ * <ul>
+ *   <li>{@link SubstraitUserDefinedAnyType} - For opaque binary UDT literals (wraps protobuf Any)
+ *   <li>{@link SubstraitUserDefinedStructType} - For structured UDT literals with accessible fields
+ * </ul>
+ *
+ * @see SubstraitUserDefinedAnyType
+ * @see SubstraitUserDefinedStructType
+ * @see io.substrait.expression.Expression.UserDefinedAny
+ * @see io.substrait.expression.Expression.UserDefinedStruct
+ */
+public class SubstraitUserDefinedType extends RelDataTypeImpl {
+
+  private final String urn;
+  private final String name;
+  private final List<io.substrait.proto.Type.Parameter> typeParameters;
+  private final boolean nullable;
+
+  public SubstraitUserDefinedType(
+      String urn,
+      String name,
+      List<io.substrait.proto.Type.Parameter> typeParameters,
+      boolean nullable) {
+    this.urn = urn;
+    this.name = name;
+    this.typeParameters =
+        typeParameters != null ? typeParameters : java.util.Collections.emptyList();
+    this.nullable = nullable;
+    computeDigest();
+  }
+
+  public String getUrn() {
+    return urn;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public List<io.substrait.proto.Type.Parameter> getTypeParameters() {
+    return typeParameters;
+  }
+
+  @Override
+  public boolean isNullable() {
+    return nullable;
+  }
+
+  @Override
+  public SqlTypeName getSqlTypeName() {
+    return SqlTypeName.OTHER;
+  }
+
+  /** Converts this Calcite type back to a Substrait {@link Type.UserDefined}. */
+  public Type.UserDefined toSubstraitType() {
+    return Type.UserDefined.builder()
+        .urn(urn)
+        .name(name)
+        .typeParameters(typeParameters)
+        .nullable(nullable)
+        .build();
+  }
+
+  /** Creates a SubstraitUserDefinedType from a Substrait Type.UserDefined. */
+  public static SubstraitUserDefinedType from(io.substrait.type.Type.UserDefined type) {
+    return new SubstraitUserDefinedType(
```

Comment:
Here you must canonize, and for doing that you need a handle to a Calcite type factory